authz/rbac_translator.go

```diff
@@ -0,0 +1,326 @@
+		return &v3matcherpb.StringMatcher{
+			MatchPattern: &v3matcherpb.StringMatcher_Suffix{Suffix: suffix},
+		}
+	}
+	return &v3matcherpb.StringMatcher{
+		MatchPattern: &v3matcherpb.StringMatcher_Exact{Exact: value},
+	}
+}
+
+func getHeaderMatcher(key, value string) *v3routepb.HeaderMatcher {
+	if value == "*" {
+		return &v3routepb.HeaderMatcher{
+			Name:                 key,
+			HeaderMatchSpecifier: &v3routepb.HeaderMatcher_PrefixMatch{PrefixMatch: ""},
+		}
+	}
+	if strings.HasSuffix(value, "*") {
+		prefix := strings.TrimSuffix(value, "*")
+		return &v3routepb.HeaderMatcher{
+			Name:                 key,
+			HeaderMatchSpecifier: &v3routepb.HeaderMatcher_PrefixMatch{PrefixMatch: prefix},
+		}
+	}
+	if strings.HasPrefix(value, "*") {
+		suffix := strings.TrimPrefix(value, "*")
+		return &v3routepb.HeaderMatcher{
+			Name:                 key,
+			HeaderMatchSpecifier: &v3routepb.HeaderMatcher_SuffixMatch{SuffixMatch: suffix},
+		}
+	}
+	return &v3routepb.HeaderMatcher{
+		Name:                 key,
+		HeaderMatchSpecifier: &v3routepb.HeaderMatcher_ExactMatch{ExactMatch: value},
+	}
+}
+
+func parsePrincipalNames(principalNames []string) []*v3rbacpb.Principal {
+	var ps []*v3rbacpb.Principal
+	for _, principalName := range principalNames {
+		newPrincipalName := &v3rbacpb.Principal{
+			Identifier: &v3rbacpb.Principal_Authenticated_{
+				Authenticated: &v3rbacpb.Principal_Authenticated{
+					PrincipalName: getStringMatcher(principalName),
+				},
+			}}
+		ps = append(ps, newPrincipalName)
+	}
+	return ps
+}
+
+func parsePeer(source peer) *v3rbacpb.Principal {
+	var and []*v3rbacpb.Principal
+	if source.Principals != nil {
+		principals := parsePrincipalNames(source.Principals)
+		if len(principals) > 0 {
+			and = append(and, principalOr(principals))
+		}
+	}
+	if len(and) > 0 {
+		return principalAnd(and)
+	}
+	return &v3rbacpb.Principal{
+		Identifier: &v3rbacpb.Principal_Any{
+			Any: true,
+		},
+	}
+}
+
+func parsePaths(paths []string) []*v3rbacpb.Permission {
+	var ps []*v3rbacpb.Permission
+	for _, path := range paths {
+		newPath := &v3rbacpb.Permission{
+			Rule: &v3rbacpb.Permission_UrlPath{
+				UrlPath: &v3matcherpb.PathMatcher{
+					Rule: &v3matcherpb.PathMatcher_Path{Path: getStringMatcher(path)}}}}
+		ps = append(ps, newPath)
+	}
+	return ps
+}
+
+func parseHeaderValues(key string, values []string) []*v3rbacpb.Permission {
+	var vs []*v3rbacpb.Permission
+	for _, value := range values {
+		newHeader := &v3rbacpb.Permission{
+			Rule: &v3rbacpb.Permission_Header{
+				Header: getHeaderMatcher(key, value)}}
+		vs = append(vs, newHeader)
+	}
+	return vs
+}
+
+// hop-by-hop headers.
+var hopHeaders = []string{
+	"Connection",
+	"Keep-Alive",
+	"Proxy-Authenticate",
+	"Proxy-Authorization",
+	"Te",
+	"Trailer",
+	"Transfer-Encoding",
+	"Upgrade",
+}
+
+func unsupportedHeader(key string) bool {
+	if strings.HasPrefix(key, ":") || strings.HasPrefix(key, "grpc-") || key == "host" || key == "Host" {
```

Comment:
Is there a reason to omit `host` and `Host` from the list above?

Reply:
Both host and Host are supported in HTTP 1.1
With the lowercase conversion, I removed the check for Host.